src/fragment-generation-utils.js

```diff
@@ -490,45 +507,59 @@ const FragmentFactory = class {
-                  .search(fragments.internal.BOUNDARY_CHARS);
-          if (newBackwardsPrefixOffset === -1) {
-            this.setBackwardsPrefixOffset(
-                this.getBackwardsPrefixSearchSpace().length);
-          } else {
-            this.setBackwardsPrefixOffset(
-                this.backwardsPrefixOffset() + 1 + newBackwardsPrefixOffset);
-          }
-        } while (
-            this.backwardsPrefixOffset() < this.getPrefixSearchSpace().length &&
-            this.getBackwardsPrefixSearchSpace()
-                    .substring(
-                        oldBackwardsPrefixOffset, this.backwardsPrefixOffset())
-                    .search(fragments.internal.NON_BOUNDARY_CHARS) == -1);
+        if (this.prefixSegments != null) {
+          this.prefixOffset = this.getNextOffsetBackwards(
+              this.prefixSegments, this.prefixOffset);
+        } else {
+          // Shift to the next boundary char, and repeat until we've added a
+          // word char.
+          const oldBackwardsPrefixOffset = this.backwardsPrefixOffset();
+          do {
+            checkTimeout();
+            const newBackwardsPrefixOffset =
+                this.getBackwardsPrefixSearchSpace()
+                    .substring(this.backwardsPrefixOffset() + 1)
+                    .search(fragments.internal.BOUNDARY_CHARS);
+            if (newBackwardsPrefixOffset === -1) {
+              this.setBackwardsPrefixOffset(
+                  this.getBackwardsPrefixSearchSpace().length);
+            } else {
+              this.setBackwardsPrefixOffset(
+                  this.backwardsPrefixOffset() + 1 + newBackwardsPrefixOffset);
+            }
+          } while (this.backwardsPrefixOffset() <
+                       this.getPrefixSearchSpace().length &&
+                   this.getBackwardsPrefixSearchSpace()
+                           .substring(
+                               oldBackwardsPrefixOffset,
+                               this.backwardsPrefixOffset())
+                           .search(fragments.internal.NON_BOUNDARY_CHARS) ==
+                       -1);
+        }
       }
-
       if (this.suffixOffset < this.getSuffixSearchSpace().length) {
-        const oldSuffixOffset = this.suffixOffset;
-        do {
-          const newSuffixOffset =
-              this.getSuffixSearchSpace()
-                  .substring(this.suffixOffset + 1)
-                  .search(fragments.internal.BOUNDARY_CHARS);
-          if (newSuffixOffset === -1) {
-            this.suffixOffset = this.getSuffixSearchSpace().length;
-          } else {
-            this.suffixOffset = this.suffixOffset + 1 + newSuffixOffset;
-          }
-        } while (this.suffixOffset < this.getSuffixSearchSpace().length &&
-                 this.getSuffixSearchSpace()
-                         .substring(oldSuffixOffset, this.suffixOffset)
-                         .search(fragments.internal.NON_BOUNDARY_CHARS) == -1);
+        if (this.suffixSegments != null) {
```

Comment:
Could it be confusing to future devs (maybe ourselves) that  suffix / prefix segments are a direct member access, but start and end segments are a getter?  As in, could we forget we need to use the getters for the start / end segments? Maybe trivial getters for prefix/suffix would mitigate that, idk

Reply:
I think that's a good call, done.